File: src/main/java/de/eugenbernwald/fretboardtrainer/model/Tone.java
package de.eugenbernwald.fretboardtrainer.model;

import java.util.Random;

public class Tone {

    private static Random random = new Random();

    private Note note;

    private Shift shift;

    private float frequency;

    private float lowerLimit;

    private float upperLimit;

    private int key;

    public Tone(Note note, Shift shift, int octave) {
        this.note = note;
        this.shift = shift;
        this.key = octave * 12 + (note.getSemitones() + shift.getValue());

        this.frequency = Tone.key2Frequency(this.key);
        this.lowerLimit = (this.frequency + key2Frequency(this.key - 1)) / 2.0f;
        this.upperLimit = (this.frequency + key2Frequency(this.key + 1)) / 2.0f;
    }

    public float getFrequency() {
        return frequency;
    }

    public float getLowerLimit() {
        return lowerLimit;
    }

    public float getUpperLimit() {
        return upperLimit;
    }

    public int getKey() {
        return key;
    }

    @Override
    public String toString() {
        return String.format("%s%s", this.note, this.shift.getSymbol());
    }

    public static float key2Frequency(int key){
        return 440.0f * (float) Math.pow(2.0f, (key - 60) / 12.0f);
    }


}

File: src/main/java/de/eugenbernwald/fretboardtrainer/model/Tone.kt
package de.eugenbernwald.fretboardtrainer.model

import java.util.*

class Tone(private val note: Note, private val shift: Shift, octave: Int) {

    val frequency: Float

    val lowerLimit: Float

    val upperLimit: Float

    val key: Int

    init {
        this.key = octave * 12 + (note.semitones + shift.value)

        this.frequency = Tone.key2Frequency(this.key)
        this.lowerLimit = (this.frequency + key2Frequency(this.key - 1)) / 2.0f
        this.upperLimit = (this.frequency + key2Frequency(this.key + 1)) / 2.0f
    }

    override fun toString(): String {
        return String.format("%s%s", this.note, this.shift.symbol)
    }

    companion object {

        private val random = Random()

        fun key2Frequency(key: Int): Float {
            return 440.0f * Math.pow(2.0, ((key - 60) / 12.0f).toDouble()).toFloat()
        }
    }


}
